parser_library/src/context/common_types.cpp

```diff
@@ -15,68 +15,19 @@
 #include "common_types.h"
 
 #include <cassert>
-#include <cctype>
 
 namespace hlasm_plugin::parser_library::context {
 
-SET_t::SET_t(context::A_t value)
-    : a_value(value)
-    , b_value(value)
-    , c_value(object_traits<C_t>::default_v())
-    , type(SET_t_enum::A_TYPE)
-{}
-
-SET_t::SET_t(context::B_t value)
-    : a_value(value)
-    , b_value(value)
-    , c_value(object_traits<C_t>::default_v())
-    , type(SET_t_enum::B_TYPE)
-{}
-
-SET_t::SET_t(context::C_t value)
-    : a_value(object_traits<A_t>::default_v())
-    , b_value(object_traits<B_t>::default_v())
-    , c_value(std::move(value))
-    , type(SET_t_enum::C_TYPE)
-{}
-
-SET_t::SET_t(const char* value)
-    : a_value(object_traits<A_t>::default_v())
-    , b_value(object_traits<B_t>::default_v())
-    , c_value(value)
-    , type(SET_t_enum::C_TYPE)
-{}
-
-SET_t::SET_t(SET_t_enum type)
-    : a_value(object_traits<A_t>::default_v())
-    , b_value(object_traits<B_t>::default_v())
-    , c_value(object_traits<C_t>::default_v())
-    , type(type)
-{}
-
-A_t& SET_t::access_a() { return a_value; }
-
-B_t& SET_t::access_b() { return b_value; }
-
-C_t& SET_t::access_c() { return c_value; }
-
-const A_t& SET_t::access_a() const { return a_value; }
-
-const B_t& SET_t::access_b() const { return b_value; }
-
-const C_t& SET_t::access_c() const { return c_value; }
-
 bool SET_t::operator==(const SET_t& r) const noexcept
 {
-    if (type != r.type)
+    if (value_type != r.value_type)
         return false;
 
-    switch (type)
+    switch (value_type)
     {
         case SET_t_enum::A_TYPE:
-            return a_value == r.a_value;
         case SET_t_enum::B_TYPE:
-            return b_value == r.b_value;
+            return a_value == r.a_value;
```

Comment:
I wonder if there still shouldn't be a special handling for `B_TYPE`?

```suggestion
            return !!a_value == !!r.a_value;
```

Reply:
I don't think this should be needed. The test at the top of the routine ensures that both objects are of `B_TYPE` and the only constructor that can create that state, takes bool, so that value in `a_value` should be either `0` or `1`.

Reply:
I don't remember now - are any conversions between types allowed? E.g. something like:

```cpp
A_TYPE a = 5;
auto b = static_cast<B_TYPE>(a);
```

Reply:
the `A_TYPE`, `B_TYPE` are just enum class members... if you look at the original constructor, `a_value` and `b_value` were always constructed in-sync. So the new implementation does the same, minus the extra `b_value` field.